### PR TITLE
Expose type synonyms in data-dependencies.

### PR DESCRIPTION
**Pull request by sofiafaro-da**
This PR converts (non-unit) type synonyms during LF conversion, and then
exposes them via data-dependencies. This is possible only when using an
LF version that supports type synonyms (LF version >= 1.7), the
type synonym isn't of unit type (because it clashes with empty
typeclasses, and the (fully applied) type synonym has kind *
(e.g. you can't define a * -> * synonym in LF).

This fixes issue #6306

changelog_begin

- [DAML Compiler] The DAML compiler will now compile type synonyms
  (``type X = Y``) into the DAR, whenever possible, and will expose
  these synonyms via data-dependencies.

changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 1]
```diff
@@ -475,6 +476,28 @@ convertSimpleRecordDef env tycon = do
...
     pure $ typeDef : [workerDef | flavour == NewtypeFlavour]
 
+convertTypeSynonym :: Env -> TyCon -> ConvertM [Definition]
+convertTypeSynonym env tycon
+    | NameIn DA_Generics _ <- GHC.tyConName tycon
+    = pure []
+
+    | envLfVersion env `supports` featureTypeSynonyms
+    , Just (params, body) <- synTyConDefn_maybe tycon
+    , isLiftedTypeKind (tyConResKind tycon)
```
Comment (cocreature) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:486]: A brief comment on this check and the one below would be great. This one seems to be to get `*`, `Constraint`, …. I’m not quite sure about the second one.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 2]
```diff
@@ -435,6 +431,11 @@ convertTypeDef env o@(ATyCon t) = withRange (convNameLoc t) $ if
     | isClassTyCon t
     -> convertClassDef env t
 
+    -- Type synonyms get expanded out during conversion (see 'convertType'), but we also
+    -- convert the synonyms we can so that we can expose them via data-dependencies.
+    | isTypeSynonymTyCon t
+    -> convertTypeSynonym env t
```
Comment (hurryabit) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:437]: Is there a technical reason for moving this further down or is it just a stylistic thing? I'm fine either way but would like to understand the former if that's the case.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 3]
```diff
@@ -475,6 +476,28 @@ convertSimpleRecordDef env tycon = do
...
+    , isLiftedTypeKind (tyConResKind tycon)
+    , not (isKindTyCon tycon)
+    = do
+        (env', tsynParams) <- bindTypeVars env params
+        tsynType <- convertType env' body
+        let tsynName = mkTypeSyn [getOccText tycon]
+        case tsynType of
+            TUnit -> pure []
+                -- ^ We avoid converting TUnit type synonyms because it
+                -- clashes with the conversion of empty typeclasses.
```
Comment (hurryabit) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:495]: Aren't we translating empty type classes into empty structural records? Isn't that different from the unit type?

File: compiler/damlc/tests/src/DA/Test/Packaging.hs [hunk 4]
```diff
@@ -1112,6 +1112,39 @@ dataDependencyTests Tools{damlc,repl,validate,davlDar,oldProjDar} = testGroup "D
...
+          writeFileUTF8 (tmpDir </> "dep" </> "daml.yaml") $ unlines
+              [ "sdk-version: " <> sdkVersion
+              , "name: dep"
+              , "version: 0.1.0"
+              , "source: ."
+              , "dependencies: [daml-prim, daml-stdlib]"
+              ]
+          writeFileUTF8 (tmpDir </> "dep" </> "Foo.daml") $ unlines
+              [ "module Foo where"
+              , "type MyInt = Int"
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs:1127]: Some more examples would be great:
1. Something with parameters.
2. Something that cannot be reconstructed to make sure things don’t explode in that case.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 1 above)
Comment (hurryabit) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:486]: Does that check that we only export synonyms of kind star?

File: compiler/damlc/tests/src/DA/Test/Packaging.hs (hunk 4 above)
Comment (hurryabit) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs:1127]: Does it make sense to put an apostrophe into the name of the synonym to make sure we're getting the mangling/unmangling right?

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 3 above)
Comment (cocreature) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:495]: LF does not allow for empty structural records.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 2 above)
Comment (sofiafaro-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:437]: I moved it down below typeclasses and constraint tuples before realizing it doesn't make a difference anyway.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 1 above)
Comment (sofiafaro-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:486]: So ... I just tested and this does indeed export constraint synonyms as well. But it also exports them without problem in data-dependencies (they work correctly, according to my tests).

So is this better to reject constraint synonyms (making the implementation more complicated), or to add a test to make sure they work (without changing the implementation)?

File: compiler/damlc/tests/src/DA/Test/Packaging.hs (hunk 4 above)
Comment (sofiafaro-da) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs]: 👍 Added!

Comment (sofiafaro-da) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs]: 👍  Done!

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 1 above)
Comment (sofiafaro-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:486]: (cc @cocreature ^^)

Comment (cocreature) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:486]: I’d say support it and add a test for it.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 3 above)
Comment (hurryabit) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:495]: I wonder why we decided to do it like that. I have no recollection... 🤔

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 1 above)
Comment (hurryabit) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:486]: Supporting and testing it sounds fine.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 2 above)
Comment (hurryabit) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:437]: Thanks for the explanation.

**Pull request merged**